activerecord/lib/active_record/future_result.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class FutureResult
+    Canceled = Class.new(ActiveRecordError)
+
+    delegate :empty?, :to_a, to: :result!
+
+    def initialize(pool, *args, **kwargs)
+      @mutex = Mutex.new
+
+      @pool = pool
+      @args = args
+      @kwargs = kwargs
+
+      @executed = false
+      @error = nil
+      @result = nil
+    end
+
+    def schedule!
+      ActiveRecord::Base.asynchronous_queries_tracker.register(self)
+      @pool.schedule_query(self)
+    end
+
+    def execute!(connection)
+      execute_query(connection)
+    end
+
+    def cancel!
+      @executed = true
+      @error = Canceled
+    end
+
+    def execute_or_skip
+      return if @executed
+
+      @pool.with_connection do |connection|
+        return unless @mutex.try_lock
+        begin
+          return if @executed
+          execute_query(connection, background: true)
+        ensure
+          @mutex.unlock
+        end
+      end
+    end
+
+    def result!
+      execute_or_wait
+      if @error
+        raise @error
+      else
+        @result
+      end
+    end
+
+    private
+      def execute_or_wait
+        return if @executed
+        if @mutex.try_lock
+          begin
+            execute_query(@pool.connection)
+          ensure
+            @mutex.unlock
+          end
+        else
+          @mutex.synchronize do
+            return if @executed
+            execute_query
+          end
+        end
+      end
+
+      def execute_query(connection, background: false)
+        @result = exec_query(connection, *@args, **@kwargs, background: background)
+      rescue => error
+        @error = error
+      ensure
+        @executed = true
```

Comment:
If the query fails to execute (like there is an exception), should `@executed` be set to true?

Reply:
I believe so yes. We mark it as executed, and when the foreground thread pops it, we re-reraise the error.

Reply:
Maybe `@pending` (and inverting the value) would be more clear? "executed" feels off in some cases (ex. when the query was cancelled).

Reply:
Sure, that makes sense.